---
title: "BOO 2025 - Example Analysis"
subtitle: "Script 2: Probe QC - Questions"
date: "`r Sys.Date()`" 
author: 
  Adam Seuren
output:
  html_document:
    code_download: true
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

***

> In this script, you will perform probe-level quality control (QC), ensuring that the final data set contains only well-expressed, relevant genes. 

***

# Setup

## Clean

As before, we perform several setup steps at the start of the script to ensure our work is reproducible and clear. 

**Exercise 1: Download the R markdown template and clean your environment.**

<details>
  <summary><strong>Hint</strong></summary>

  Use the code dropdown menu in the top right corner of this document.

</details>


```{r clean}
# The rm function clears the environment from all objects. 
rm(list=ls())
```

***

## Load packages

Here, we introduce a second package:

* `readxl` allows you to read in Excel files, such as the probe manifest

**Exercise 2: Load `tidyverse` and `readxl` into your R environment**

<details>
  <summary><strong>Hint</strong></summary>

  You may need to install packages before you load them for the first time with the `install.packages()` function.

</details>

```{r load-packages, warning=F, message=F}
# Loading the needed packages (loaded more than specified, because more packages were needed)
library(readxl)
library(tidyverse)
library(dplyr)
library(sessioninfo)
library(ggplot2)

```

***

## Set variables

As before, we set directories and commonly used variables at the start of the script to make it easier to adjust anything later.

**Exercise 3: Create the following objects:**

* `root_dir` - the root directory for your project
* `count_path` - the location of the count data
* `metadata_path` - the location of the metadata
* `manifest path` - the probe manifest for TempO-Seq probes

* `count_store` - where you want to save the count data after QC
* `cpm_store` - where you want to store the CPM data
* `metadata_store` - where you want to save the metadata
* `probedata_store` - where to store the updated manifest

```{r set-variables}
# Setting the root directory to the folder for this analysis, setting the paths to the countdata, metadata, and probe manifest.
root_dir <- "C:/Users/mmpse/Documents/GitHub/BOO-Adam/QC/02_Probe_QC/"
count_path <- "Input/countData.Rdata"
metadata_path <- "Input/metaData.Rdata"
manifest_path <- "Input/Probe_Manifest.xlsx"

# Setting the paths to the place where future objects will be saved

count_store <- "Output/countData.Rdata"
cpm_store <- "Output/cpmData.Rdata"
metadata_store <- "Output/metaData.Rdata"
probedata_store <- "Output/probeManifest.Rdata"
```

***

## Load data

For the probe QC, we will use the count data, metadata, and probe manifest.

**Exercise 4: Load required data into the environment:**

<details>
  <summary><strong>Hint</strong></summary>

  The `read_xlsx()` function from the `readxl` package can be used to read in Excel files.

</details>

```{r load-data, warning=F, message=F}
# count data, metadata and probe manifest are loaded.
load(paste0(root_dir, count_path))
load(paste0(root_dir, metadata_path))
manifest <- read_xlsx(paste0(root_dir, manifest_path))

# original count data is stored for potential use later on
countData_original <- countData
```

***

# Annotation

## Motivation

In TempO-Seq, each gene is detected using a unique probe sequence. In order to ensure accuracy in our gene annotations and downstream analyses, we need to confirm that these probes align to the locations on the genome that we expect them to.

To do this, we used [NCBI BLAST (Basic Local Alignment Search Tool)](https://blast.ncbi.nlm.nih.gov/Blast.cgi), a tool that allows us to compare the probe sequences against a reference genome database. By uploading our probe sequences to BLAST, we could:

1. Verify that each probe still uniquely matches the gene it was designed for,
2. Check for any outdated or incorrect annotations, especially since the genome has been updated since the probes were designed, and
3. Update our manifest file (the reference file we use to link probe names to gene IDs in this script).

**Exercise 5: What problems could arise in your project if you did not update the probe annotations?**

The probe sequences might have changed, which means that the sequence may not refer to the gene anymore, or that it might refer to multiple genes at once.

***

## Update

The manifest provided during this project contains updated information about the probes. 

**Exercise 6: Describe the columns in the probe manifest**

```{r manifest-inspect}
# probe_name_old: probe name from before update
# probe_ID: probe identificator
# probe_sequence: DNA sequence of the probe
# probe_coord: location on the genome where the probe is matched to
# gene_symbol: gene the probe belongs to
# entrez_id, ensembl_id, hgnc_id: several identificators for the gene
# gene_symbol_old, entrez_id_old: gene symbol and entrez id from before update
# percentage_match: percentage with which the probe matches to the gene sequence
str(manifest)

```

***

**Exercise 7: Check that we have annotation information for all of the probes in `countData`:**

```{r anno-check}
#

setequal(rownames(countData), manifest$probe_name_old)

```
The command returns TRUE, which means that there is annotation information for all of the probes in countData.
***

Sometimes, the probe sequence could no longer be aligned to a specific gene in the reference genome. In this instance, the `gene_symbol` column will be NA.

**Exercise 8: How many probes could not be aligned in the updated manifest? What should be do with these probes and why?**

```{r not-aligned}
#
NA_gene_symbol<- sum(is.na(manifest$gene_symbol))

NA_gene_symbol
```
603 values in the gene_symbol column are NA. These probes should be omitted from the countdata, as they do not refer to any gene and therefore do not provide any useful information.
***

**Exercise 9: How many probes have had their annotation updated by the manifest?**

<details>
  <summary><strong>Hint</strong></summary>

  The annotation is updated if the old and updated `gene_symbol` do not match.

</details>

```{r update-anno}
# Counting the amount of genes that still have the same symbol as before the update

manifest %>%
  count(gene_symbol == gene_symbol_old)

```
There are 608 probes that have had their annotation updated by the manifest.
***

**Exercise 10: Create `flag_realignment`, a new column in `probeManifest` which describes whether a probe should be removed, renamed, or kept unchanged:**

<details>
  <summary><strong>Hint</strong></summary>

  Try using `case_when()` inside the `mutate()` function.

</details>

```{r flag-realign}
# Adding an empty column called flag_realignment
manifest %>%
  add_column(flag_realignment = NA)

# flag_realignment column now describes whether a probe should be removed, renamed or kept unchanged.
manifest <- manifest %>%
  mutate(flag_realignment =  
           # when the new gene symbol is equal to the old one, write "unchanged"
    case_when(gene_symbol_old == gene_symbol ~ "unchanged", 
              # when the new gene symbol is not equal to the old one, write "renamed"
              gene_symbol_old != gene_symbol ~ "renamed", 
              # when the new gene symbol is NA, write "removed"
              is.na(gene_symbol) ~ "removed"))

# checking if the flag realignment went well. If the realignment went well, this should say: removed: 603, renamed: 608, unchanged: 21322, NA: 0
table(manifest$flag_realignment, useNA = "always")

```

***

If probes could not be aligned, they should be removed from the `countData`.

**Exercise 11: Subset `countData` to keep only the probes with correct annotations.**

```{r subset-realign}
# subsetted countData to only show the probes which do not have the tag "removed" in the flag_realignment column of manifest_new
countData <- subset(countData, row.names(countData) %in% manifest$probe_name_old & manifest$flag_realignment != "removed")



```

***

As well as removing probes that could not be annotated, we also want to update the names of probes where the gene name has changed.

**Exercise 12: Create `probe_name`, a new column in `probeManifest` that contains the updated probe names, and adjust the `countData` row names accordingly:**

<details>
  <summary><strong>Hint</strong></summary>

  The new probe name should be a combination of `gene_symbol` and `probe_ID`.

</details>

```{r probe-name}
# The mutate verb adds a probe_name column. The new probe names are gene_symbol and probe_ID, separated by an underscore.
manifest <- manifest %>%
  mutate(probe_name = paste(gene_symbol, probe_ID, sep = "_"))

countData <- countData %>%
  # turning the row names (probes) back into a column
   rownames_to_column(var = "probe_name_old") %>% 
  # merging the countData with manifest, to the left of probe_name_old
  left_join(y = manifest, by= "probe_name_old") %>%
  # selecting only the probe_name and sample_ID columns, these are the two columns of the manifest that will be used in the countData
  select(probe_name, all_of(metaData$sample_ID)) %>%
  # changing the column probe_name to rownames of the countData: probe_name_old has now been swapped for the new probe_name.
  column_to_rownames(var="probe_name")

```

***

It is good practice to check the dimensions of objects when making such adjustments using the `dim()` function. 

**Exercise 13: Check the dimensions of the objects in your environments to ensure nothing went wrong:**

```{r dim-check}
# checking dimensions of the manifest (should be 22533/13 for manifest and 21930 (=22533-603) for countData, 66/29 for metaData) 
dim(manifest)
dim(countData)
dim(metaData)
```

***

# Normalization

## CPM

Counts Per Million (CPM) normalization is a method that accounts for differences in sequencing depth between samples. It scales raw read counts (as in `countData`) so that the total counts per sample sum to one million. This allows for comparison of gene expression levels across samples.

**Exercise 14: Calculate CPM values for our data and store them in a `cpmData` object:**

```{r cpm-norm}
# calculating the counts per million and putting them in a dataframe
cpmData <- data.frame(
  # the apply function is used to apply the calculation to the rows ("2") of the countData
  apply(countData, 2, 
        # cpm = (counts / sum of all counts in the row) * 1 million
        function(x){x/sum(x)*1000000}))
```

***

# Probe relevance

## Highly expressed genes

By inspecting the most highly expressed genes in a dataset, you can build confidence in the data and get an idea of processes with particular relevance to your project.

**Exercise 15: Add the columns `total_cpm` and `average_cpm` to the probe manifest and describe a few of the most highly expressed genes. Do they make biological sense?**

<details>
  <summary><strong>Hint</strong></summary>

  You can look up gene functions using [NCBI](https://www.ncbi.nlm.nih.gov/gene/) or [GeneCards](https://www.genecards.org/).

</details>

```{r cpm-summarize}
# Create a data frame of the total CPM (is the sums of each row in cpmData, rounded to 3 decimals) for each probe
manifest <- data.frame(total_cpm = round(rowSums(cpmData), 3)) %>% 
  # Put the rownames from total_cpm into a column named probe_name
  rownames_to_column(var="probe_name") %>%
  # Calculate the average CPM across samples for each probe (= total cpm divided by the number of columns (ncol) of cpmData (each column in cpmData is a sample)). This is, again, rounded to 3 decimals. 
  mutate(avg_cpm = round(total_cpm / ncol(cpmData), 3)) %>% 
  # Add the avg_cpm and total_cpm column to manifest, to the right of probe_name
  right_join(manifest, by="probe_name")

# showing the probes in descending order of expression level, so the highest expressed probe is at the top.
highest_expressed_probes <- manifest %>%
  arrange(desc(avg_cpm))
highest_expressed_probes
```

HSPB1: encodes a heat shock protein, which is expressed in response to stress. This makes sense, as (some of the) cells have been treated with a toxic chemical.
APOA1: encodes a plasma protein, is expressed more highly in the liver. 
APOC3: encodes a metabolism protein, this makes biological sense as the liver plays a big role in metabolism.
ALB: encodes albumin. This protein is highly expressed in the blood.
ORM1: encodes a plasma protein that is mostly expressed in reaction to acute inflammation. This could be an adverse reaction to the toxic chemicals.

***

## Lowly expressed genes

Although these highly expressed genes are interesting and informative, there are also many lowly expressed genes in our data. When genes have very low expression levels, small absolute changes can be represented by very high fold changes. We want to remove these lowly expressed and noisy genes from our data with the relevance filter.

**Exercise 16: Visualize the distribution of log2(average CPM + 1) across your data. Can you identify these lowly expressed genes?**

<details>
  <summary><strong>Hint</strong></summary>

  Genes with an average CPM below 1 are generally considered to be lowly expressed.

</details>

```{r cpm-plot}
#
manifest %>%
  # subsetting the manifest so the unassigned probes are not included
  filter(flag_realignment != "removed") %>%
  # mutating avg_cpm to log2(avg_cpm+1), so it can be plotted
  mutate(log2avg_cpm = log2(avg_cpm+1)) %>% 
  ggplot() +
  # plotting log2avg_cpm as a histogram, with light blue filled bars and a binwidth of 0.2
  geom_histogram(aes(x=log2avg_cpm), fill='lightblue', color='black', binwidth = 0.2) +
  # creating a red, dashed vertical line at x = 1; this is the threshold for lowly expressed genes. 
  geom_vline(aes(xintercept=1), color='red', linetype='dashed') +
  # labelling the x-axis so that it is clear what the plot means
  xlab('log2 of the average CPM (+1) per sample for each probe') +
  # The theme makes the plot look cleaner.
  theme_bw()


```
The genes that are plotted on the left side of the red line can be considered lowly expressed.
***

**Exercise 17: Why did you use a log2 transformation here and why was it necessary to add 1 to the CPM values?**

The log2 transformation is done to improve the normal distribution; normally, gene expression is skewed because expression can not go below 0.

It is necessary to add 1 to the CPM values, so that no log of 0 will be taken. This way, all values are included, as ggplot will otherwise remove them.

***

**Exercise 18: How many probes have no expression measured in any of the samples?**

```{r no-expression}
# Probes with no expression have a total_cpm of 0. These are now filtered from the manifest.
manifest %>%
  filter(total_cpm == 0)
```
406 probes have no measured expression.

***

## Relevance

We want to keep probes in our data if they are relevant for any of our experiments. Here, we define relevance as having a CPM above 1 for at least two thirds of the replicates. 

**Exercise 19: How many experimental conditions do we have in this data and how many replicates are there for each?**

```{r replicate-number}
#
metaData %>% 
  group_by(mean_ID) %>% 
  # summarize the number of reps for each condition.
  #.groups = "drop" removes the grouping done by group_by, this prevents errors from happening later (if this dataframe would be used later on). 
  summarize(n_reps = n(), .groups = "drop") %>% 
  # make sure there are no duplicate conditions. 
  distinct()
```
There are 18 conditions in total. The treated conditions (12) all have 3 replicates, and the DMSO conditions (4) all have 6 replicates.

***

**Exercise 20: Loop through these experimental conditions, calculating for each the percentage of replicates where each probe is above the 1 CPM threshold:**

<details>
  <summary><strong>Hint</strong></summary>

  First create a vector of the experimental conditions to use in a `for()` loop.

</details>

```{r experiment-loop}
# Creating a vector named loop, containing the experimental conditions
loop <- unique(metaData$mean_ID)
loop
```

```{r}
# Creating the dataframe relevanceFilter, which contains only the rownames of cpmdata
relevanceFilter <- data.frame(probe_name = rownames(cpmData))

# For each unique mean ID (= each unique condition)
for(i in loop){
  # Subset metadata to include only samples with that mean ID (so, for each loop, the samples of one unique condition are selected)
  set <- metaData %>% 
    filter(mean_ID == i)
  
  # Subset the CPM data to include only those samples (so you have the cpm data for all repetitions of the condition)
  cpm.set <- cpmData[ , set$sample_ID]
  
  # Count the percentage of values for each probe that are above 1 (CPM above 1 is the threshold for relevance. The percentage values indicate the percentage of replicates in which the value for each probe is above 1)
  df.set <- data.frame(round(100 * rowSums(cpm.set > 1) / ncol(cpm.set), 2)) %>% 
    rownames_to_column(var = "probe_name")
  
  # Set the column name to the current mean ID
  colnames(df.set) <- c("probe_name", i)
  
  # Merge with the relevance filter output thus far (so that all values will appear in the same dataframe)
  relevanceFilter <- left_join(relevanceFilter, df.set, by="probe_name")
}

# Turn the probe name column back into row names
relevanceFilter <- relevanceFilter %>% 
  column_to_rownames(var="probe_name")

# Show the first 6 rows of the output
head(relevanceFilter)

```


***

**Exercise 21: Add a column `n_relevant` to the relevance filter output which describes for how many experiments each probe is relevant. Merge the output with the probe manifest:**

<details>
  <summary><strong>Hint</strong></summary>

  The `rowSums()` function may come in handy :)

</details>

```{r count-relevant}
relevanceFilter <- relevanceFilter %>% 
  # Create a variable counting the number of rows with a value above 66 -> a probe is relevant if it has a value of >1 in at least two thirds of the data (so at least 66% of the replications). 
  mutate(n_relevant = rowSums(relevanceFilter >= 66)) %>% 
  # Move this column to the first place (left) in the data frame
  select(n_relevant, everything()) %>% 
  # Convert row names to a new column called probe_name
  rownames_to_column(var="probe_name")

# Add the relevanceFilter to the manifest using this new probe_name column
manifest <- left_join(manifest, relevanceFilter, by="probe_name")

# Summarize the number of relevant experiments (minimum, 1st and 3rd quantile, median, mean, maximum) for each probe 
summary(manifest$n_relevant)

```

***

## Flag

We want to flag the lowly expressed probes in the manifest and remove them from our `countData`. These noisy, unreliable probes should not be included in our downstream analysis.

**Exercise 22: Flag any probes that do not pass the relevance filter for at least one experimental condition and remove them from `countData`:**

```{r flag-relevance}
# A probe is relevant if it passes the threshold (cpm >1 in at least 2/3 of the repetitions) for at least 1 experimental condition. TRUE indicates a relevant probe, FALSE indicates an irrelevant probe.
manifest <- manifest %>% 
  mutate(flag_relevant = ifelse(n_relevant >= 1, T, F))

# Put the relevance into a table
table(manifest$flag_relevant, useNA="always")
```
There are 7248 irrelevant probes in the manifest.

```{r}
# Put the probes that pass the filter into a vector
kept <- (manifest %>% 
  filter(flag_relevant == T))$probe_name

# Check the dimensions of the count data to make sure they are still correct (should be 21930/66)
dim(countData)
```
```{r}
# Subset the count data so that it contains only the relevant probes
countData <- countData[kept, ]

# Check dimensions again following the subset (number of rows should be equal to the number of relevant probes, which is 14646)
dim(countData)
```
Now, countData only contains relevant probes, that are well-expressed in at least 2/3 of the repetitions of at least one of the conditions.
***

**Exercise 23: Visualize and interpret the distribution of `n_relevant`:**

```{r relevant-plot}
# manifest is filtered so only assigned probes are used in the plot
manifest %>% 
  filter(flag_realignment != "Probe removed") %>% 
  ggplot() +
  # A histogram is made with the number of relevant probes on the x-axis.
  geom_histogram(aes(x = n_relevant,
                     # Bins for relevant and irrelevant probes will be filled with a different color
                     fill = flag_relevant),
                 color = "grey20",
                 binwidth = 1) +
  # setting axis labels and a figure title
  xlab(paste0("Number of conditions where a probe reaches 1 CPM in at least 66% reps")) +
  ylab("Number of probes") + 
  ggtitle("Relevance distribution") + 
  # theme makes the plot look cleaner
  theme_bw() +
  # Remove the legend
  theme(legend.position = "none")
```
The histogram has high bins on both outer ends of the plot. This means that probes are, most of the time, relevant in either all or none of the conditions.
```{r}
# filtering the manifest so it shows only the probes that are relevant in all of the conditions (where n_relevant is at its maximum level). NA values are removed.
manifest %>% 
  filter(n_relevant == max(n_relevant, na.rm=T))
```
11,112 probes are relevant in all of the samples. This is roughly half of the total amount of probes, and (as was also seen in the histogram), a large part of the total relevant probes.
***

**Exercise 24: Check the distribution of log2(CPM + 1) values coloured by the relevant filter flag. Has the peak of lowly expressed genes been reduced?**

```{r dist-check}
manifest %>% 
  # unaligned probes are filtered out
  filter(flag_realignment != "Probe removed") %>% 
  # add a column with the log2(cpm + 1) of the average cpm
  mutate(log2avg_cpm = log2(avg_cpm + 1)) %>% 
  ggplot() +
  # create a histogram with the log2avg_cpm on the x-axis 
  geom_histogram(aes(x = log2avg_cpm, 
                     # bins are color-filled by relevance
                     fill = flag_relevant), 
                 color = "black", binwidth = 0.2) +
  # label is added on the x-axis, plot title is added
  xlab("log2 of the average CPM (+1) per sample for each probe") +
  ggtitle("Distribution of log2 (CPM + 1) values colored by relevance") +
  # the theme makes the plot look cleaner
  theme_bw() +
  # remove the legend from the plot
  theme(legend.position='none')
```
As can be seen from the color of the bins, the peak of lowly expressed genes has been removed by the relevance filter.
***

# Gene-level data

## Sum

Multiple probes from TempO-Seq may interrogate the same gene. Since all of our downstream analyses will be performed at the gene level, we sum probes interrogating the same gene to create gene-level data. 

**Exercise 25: Sum the counts of probes interrogating the same gene and store the gene-level data:**

<details>
  <summary><strong>Hint</strong></summary>

  Use the `separate` command on the probe names to create a gene name variable easily in the `countData`.

</details>

```{r gene-level}
countData <- countData %>% 
  # change the row names of countData into a column called probe_name
  rownames_to_column(var = "probe_name") %>%  
  # reduce the probe name to just the gene symbol, by removing the probe ID (making it NA) and separator.
  separate(probe_name, 
           into=c("gene_symbol", NA), 
           sep = "_") %>% 
  # group the data by gene symbol
  group_by(gene_symbol) %>% 
  # add the values in each row to each other, so that the counts of each gene are taken together. 
  summarise(across(where(is.numeric), sum), 
            # remove the grouping to avoid future errors
            .groups = "drop") %>% 
  # turn the column gene_symbol into row names
  column_to_rownames(var = "gene_symbol")

# show the first 10 rows of countData
countData[1:10, ]
```
Raw counts for probes referring to the same gene have now been added together.
***

**Exercise 26: Re-apply CPM normalization on this gene level data**

```{r reapply-cpm}
cpmData <- data.frame(
  # for each row, calculate cpm (cpm = (counts / sum of all counts in the column) * 1 million)
  apply(countData, 2, function(x){x/sum(x)*1E6}))

cpmData[1:10, ]
```

***

# Advanced questions

If you would like a bit more of a challenge, here are a few extra questions relating to the two probe QC steps above. However, you can also skip these, save your data, and move on to the sample QC.

## Probe realignment

Some of the probes that could not be aligned to the most recent reference genome interrogate genes that are still present in the data. 

**Exercise 27: For genes interrogated by more than 5 probes where at least one of these was removed during realignment, investigate whether the removed probe measured more or fewer counts than the ones remaining in the data.**

<details>
  <summary><strong>Hint</strong></summary>

  In order to answer these questions at the end of the script (after all the filters), you may want to save a `countData_original` object on loading count data into the environment.

</details>

```{r advanced-realign}
#

```

***

## Relevance filter

Some probes that were kept in the data had very low average expression levels, but were deemed relevant for one of the experimental conditions.

**Exercise 28: Take a look at the most extreme examples of such probes. What conditions are responsible for them being kept in the data?**

```{r advanced-relevance}
#

```

***

# Save

**Exercise 29: Save the following objects:**

* `probeManifest` containing information on probes, their expression levels, and the results of the different filters
* `countData` containing the filtered count data for probes that should be used in downstream analysis
* `cpmData` containing the CPM normalized counts for relevant, well-expressed probes

```{r save-data}
# saving the manifest, countData and cpmData to the output file (via the paths defined earlier)
save(manifest, file=paste0(root_dir, probedata_store))
save(countData, file=paste0(root_dir, count_store))
save(cpmData, file=paste0(root_dir, cpm_store))
```

***

# Session Info

**Exercise 30: Print your session info at the end of the script to ensure reproducibility:**

```{r session-info}
#
devtools::session_info()
```

***

**Exercise 31: Knit this script to produce either a `PDF` or `HTML`file and push it to your GitHub repository.**

That is the end of the Probe QC. Example answers will be available from the `BOO_template` GitHub on Tuesday. 

Next, please move on to the probe QC using `03_SampleQC_Outline.Rmd`.

***

