---
title: "BOO 2025"
subtitle: "Functional enrichment"
date: "`r Sys.Date()`" 
author: 
  Adam Seuren
output:
  html_document:
    code_download: true
    theme: united
    lowlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

***

> In this script, you will biologically interpret your differentially expressed genes.

***

# Setup

## Clean

As before, we perform several setup steps at the start of the script to ensure our work is reproducible and clear. 

**Exercise 1: Download the R markdown template, clean your environment, and set the folhighing variables:**

* `root_dir` - project folder
* `low_store` - location of DEG results for the low toxicity compound
* `high_store` - location of DEG results for the high toxicity compound

```{r clean}
rm(list=ls())

root_dir <- "C:/Users/mmpse/Documents/GitHub/BOO-Adam/ORA/"

low_fetch <- "Input/DEGres_low.Rdata"
moderate_fetch <- "Input/DEGres_moderate.Rdata"

low_store <- "Output/DEGres_low.Rdata"
moderate_store <- "Output/DEGres_moderate.Rdata"
```

***

## Packages

In this script, we introduce a new package: `enrichR`, a package for gene set enrichment analysis that identifies biological pathways, gene ontology terms, and other functional annotations significantly associated with a given list of genes.

**Exercise 2: Load `enrichR` and `tidyverse` into your R environment:**

```{r load-packages, warning=F, message=F}
library(tidyverse)
library(enrichR)
```

***

## Load data

**Exercise 3: Load the DEG results R environment:**

```{r load-data, warning=FALSE, message=FALSE}
load(paste0(root_dir, moderate_fetch))
load(paste0(root_dir, low_fetch))
```

***

# ORA

## Databases

When performing any type of gene set analysis, you need to specify the gene sets of interest. This is one of the pitfalls but also benefits of this type of analysis.

**Exercise 4: Use `listEnrichrDbs()` and look at the gene set databases available. Select a few of interest and describe them:**

<details>
  <summary><strong>Hint</strong></summary>

  Sometimes a specific database has different version for different years. Try to select the most recent ones using `grepl`.

</details>

```{r}
listEnrichrDbs() %>% 
  # Preference is given to versions from 2020 or newer
  filter(grepl("202", libraryName))
```

* `MSigDB_Hallmark_2020`: gene sets curated by the Broad Institute from the Molecular Signatures Database (MSigDB), repesenting coherent biological states or processes aimed to reduce redundancy in gene set databases.
* `TG_GATES_2020`: derived from the Toxicogenomics Project-Genomics Assisted Toxicity Evaluation System (TG-GATEs), this database links gene expression changes to specific chemical exposures in rodent and human hepatocytes, providing mechanistic insights into toxicity and drug safety.
* `GO_Biological_Process_2025`, `GO_Cellular_Component_2025`, and `GO_Molecular_Function_2025` are part of the Gene Ontology (GO) Consortium, and capture structured biological processes (e.g. inflammatory response), subcellular locations (e.g. mitochondrial matrix), and molecular activities (e.g. ATP binding), respectively.
* `WikiPathways_2024_Human` is an open-source, community-curated collection of human biological pathways hosted by the WikiPathways project, aimed at enhancing transparency and collaboration in pathway annotation across a range of processes and diseases. 
* `Reactome_Pathways_2024` is a peer-reviewed and manually curated database of human pathways developed by the Reactome Consortium, emphasizing detailed molecular interactions and regulatory mechanisms across signalling, metabolism, and disease.
* `KEGG_2021_Human` is developed by the Kyoto Encyclopedia of Genes and Genomes (KEGG) and contains human pathway maps integrating genomic, chemical, and functional data to support systems-level understanding of cellular and organismal functions.

***

## DEG List

In order to run ORA you need a list of genes of interest (your DEGs).

**Exercise 5: Make a DEG list for each of the two compounds. How many DEGs do you have for each?**

<details>
  <summary><strong>Hint</strong></summary>

  If you compound shows a dose response, you may want to select genes at a particular concentration of interest.

</details>

```{r}
# Significant DEGs are filtered out of the data. Only their gene symbols are kept.
deg_moderate <- unique((res_moderate %>% 
               filter(sig == T))$gene_symbol)

deg_low <- unique((res_low %>% 
               filter(sig == T))$gene_symbol)

deg_moderate
deg_low
```

There are 127 DEGs for the moderate toxicity compound and 55 for the low toxicity one.

***

## Enrichr

The `enrichr` function performs over-representation analysis (ORA) using a list of genes in each of the databases specified by the `databases` argument, returning a list.

**Exercise 6: Run `enrichr` using the selected databases and inspect the output for each compound. Are there any pathways you expected?:**

<details>
  <summary><strong>Hint</strong></summary>

  You can use `bind_rows()` to convert lists into data frames and make inspecting the output easier.

</details>

```{r}
# The DEG list from cells treated with a moderately toxic compound is enriched with four databases.
ora_moderate <- enrichr(deg_moderate,
                       c("GO_Biological_Process_2025",
                         "GO_Molecular_Function_2025",
                         "Reactome_Pathways_2024",
                         "KEGG_2021_Human"))

# The list of enriched DEGs is converted into a dataframe
ora_moderate <- bind_rows(ora_moderate)

# Only the significantly changed pathways are investigated. 
ora_moderate %>% filter(Adjusted.P.value <= 0.05)
```
The top 5 enriched terms are:
- Cytoplasmic Translation: translation of mRNA into proteins.  -> could be a reaction to toxicity?
- Ribosomal Small Subunit Biogenesis: makes sense in combination with cytoplasmic translation; ribosomes are needed for the translation.
- Cellular response to low-density lipoprotein particle stimulus: triazine is a low-density lipoparticle. This is clearly a response to the chemical, though it does not directly indicate toxicity.
- Macromolecule biosynthetic process: synthesis of a macromolecule. Does not seem to be directly related to triazine.
- Gene expression: again, a logical combination with both cytoplasmic translation and ribosomal small subunit biogenesis. Overrepresentation of these three enriched terms could be a reaction to triazine entering the system.

***

```{r}
ora_low <- enrichr(deg_low,
                       c("GO_Biological_Process_2025",
                         "GO_Molecular_Function_2025",
                         "Reactome_Pathways_2024",
                         "KEGG_2021_Human"))

ora_low <- bind_rows(ora_low)

ora_low %>% filter(Adjusted.P.value <= 0.05)
```
Top 5 enriched terms:
- Retinoid metabolic process: metabolism of retinoids - derivatives of vitamin A
- Aminoglycoside antibiotic metabolic process:
- Daunorubucin metabolic process:
- Prostaglandin metabolic process:
- Polyketide metabolic process:

These, and most other high-ranking enriched terms, describe metabolic processes. The reaction of the cell to propylene carbonate might have something to do with metabolism. -> search literature
***

## Plots

## Bar plots

Using `geom_col`, we can visualize the results for each of the compounds.

**Exercise 7: Take a look at the top enriched terms for each compound:**

```{r}
ora_moderate %>% 
  # The 10 enriched terms with the lowest adjusted p-values are shown, ordered by ascending adjusted p-value.
  arrange(Adjusted.P.value, P.value) %>% 
  head(10) %>%
  # The p-value is log-adjusted
  mutate(log10p = -log10(P.value)) %>% 
  # Size of the total pathway is separated from the number of genes found in the pathway, so that the bars can be colored by geneset size
  separate(Overlap, 
           sep="/", 
           into=c("Genes_found", "Size"),
           remove=F) %>% 
  mutate(Size = as.numeric(Size)) %>% 
  ggplot(aes(y=reorder(Term, log10p))) + 
  geom_col(aes(x=log10p, fill = Size),
           color='black') +
  ylab('') + xlab('-log10p') + ggtitle("Top ten enriched terms") +
  # Y-axis is labelled with the enriched terms
  scale_y_discrete(labels = function(x) str_wrap(x, width = 30)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=7))
```
-> find link of these to potential toxicity
- SRP-dependent cotranslational protein targeting to membrane
- nonsense-mediated decay enhanced by exon junction complex, nonsense-mediated decay: do not seem to be directly linked to toxicity/ adverse reactions.
- regulation of expression of SLITs and ROBOs, signaling by ROBO receptors: SLIT/ROBO pathway is involved in cell migration, cell death, angiogenesis. Mainly active during development, but also in tumors.
- Eukaryotic translation termination
- Cellular responses to stress: presumably an adverse reaction to triazine
- Cellular responses to stimuli: presumably a reaction to triazine
- Sars-CoV-2 infection, Sars-CoV-2-host interactions: Does not seem to be a direct reaction to triazine


```{r}
ora_low %>% 
  arrange(Adjusted.P.value, P.value) %>% 
  head(10) %>%
  mutate(log10p = -log10(P.value)) %>% 
  separate(Overlap, 
           sep="/", 
           into=c("Genes_found", "Size"),
           remove=F) %>% 
  mutate(Size = as.numeric(Size),
         Genes = str_wrap(Genes, width = 10)) %>% 
  ggplot(aes(y=reorder(Term, log10p))) + 
  geom_col(aes(x=log10p, fill = Size),
           color='black') +
  ylab('') + xlab('-log10p') + ggtitle("Top ten enriched terms: ") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 30)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=7))
```

***

## Upset plot

You may also want to visualize the top enriched terms for both compounds together.

**Exercise: Use `geom_tile()` to do this:**

First we select terms to use in the plot (top 10 for each compound):

```{r}
top_terms <- c((ora_moderate %>% head(10))$Term, (ora_low %>% head(10))$Term)
top_terms
```

***

Then we bind the results together, making a new variable specifying if the compound is high or low toxicity:

```{r}
ora_moderate$Toxicity <- "High"
ora_low$Toxicity <- "Low"

ora_both <- rbind(ora_moderate, ora_low)
ora_both
```

Then, we can make the plot:

```{r}
ora_both %>% 
  filter(Term %in% top_terms) %>% 
  ggplot(aes(y=Term, x=Toxicity)) +
  geom_tile(color="grey50", fill="grey90", linewidth = 0.3) +
  geom_point(aes(fill=Adjusted.P.value, size=Odds.Ratio), shape=21, color='black') +
  ylab("") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50),
                   expand=c(0,0)) +
  scale_x_discrete(expand=c(0,0)) +
  scale_fill_gradientn(colors = c('dodgerblue', 'black')) +
  ggtitle('Top enriched terms in both compounds:') +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size=6),
    panel.background = element_rect(fill='white', color='grey50'),
    panel.grid = element_blank()
  )
```



***

# Session Info

```{r session-info}
devtools::session_info()
```

***

