---
title: "BOO 2025"
subtitle: "Differential expression"
date: "`r Sys.Date()`" 
author: 
  Adam Seuren
output:
  html_document:
    code_download: true
    theme: united
    lowlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

***

> In this script, you will identify differentially expressed genes (DEGs) and visualize your results.

***

# Setup

## Clean

As before, we perform several setup steps at the start of the script to ensure our work is reproducible and clear. 

**Exercise 1: Download the R markdown template, clean your environment, and set the following variables:**

* `root_dir` - project folder
* `count_path` - location of the `countData` object
* `metadata_path` - location of the `metaData` object
* `low_store` - location to store DEG results for the low toxicity compound
* `high_store` - location to store DEG results for the high toxicity compound

```{r clean}
rm(list = ls())

root_dir <- "C:/Users/mmpse/Documents/GitHub/BOO-Adam/DEG/"
count_path <- "Input/countData.Rdata"
metadata_path <- "Input/metaData.Rdata"

low_store <- "Output/DEGres_low.Rdata"
moderate_store <- "Output/DEGres_moderate.Rdata"
```

***

## Packages

In this script, we introduce a new package: `DESeq2`, a package that analyses count-based sequencing data, such as TempO-Seq, to identify differentially expressed genes between experimental conditions.

**Exercise 2: Load `DESeq2`, `ggrepel`, and `tidyverse` into your R environment:**

```{r load-packages, warning=F, message=F}
#
library(tidyverse)
library(ggrepel)
library(DESeq2)
```

***

## Load data

**Exercise 3: Load the count and metadata into your R environment:**

```{r load-data}
#
load(paste0(root_dir, count_path))
load(paste0(root_dir, metadata_path))

```

***

# DEG Analysis 

## Mean ID

In order to perform DEG analysis we need to ensure we have the correct `mean_ID` and `mean_ID_control` variables. These can be used to tell DESeq2 which samples to compare:

* `mean_ID` is a group identifier, describing which treatment condition each sample belongs to
* `mean_ID_control` is a control identifier, describing which control samples a group should be compared to

In the case of our data we will have:

* A `mean_ID` level for each experimental condition (6 concentrations for 2 compounds plus two plate-specific DMSO groups)
* No `mean_ID_control` for the DMSO controls, but the appropriate DMSO control specified for the treated samples

**Exercise 4: Identify these variables in your `metaData`:**

```{r inspect}
# Selecting the experimental conditions and their control conditions from the metadata
metaData %>% 
  select(mean_ID, mean_ID_control) %>% 
  # Filtering out the NA values (= untreated cells) from the selected data, do not show duplicate values
  filter(!is.na(mean_ID_control)) %>% 
  distinct()
```

***

## Contrast

DESeq2 uses an object called a contrast, which details the comparisons that will be run. We often want to save several comparisons together in one object, such as all results for each compound together.

**Exercise 5: Create a `contrast` data frame, including a variable `file_ID` which describes which file each comparison should be saved in:**

```{r contrast}
# Filtering out NA values (DMSO-treated samples) in the mean_ID_control column, selecting the same columns from the metadata as in the previous chunks, changing the compound_ID column name to file_ID
contrast <- metaData %>% 
  filter(!is.na(mean_ID_control)) %>% 
  select(mean_ID, mean_ID_control, file_ID = compound_ID) %>% 
  distinct

contrast

```
This dataframe shows which control condition belongs to which treatment condition.
***

## Subsets

We want to perform DEG analysis for each concentration of the low and high toxicity compounds.

**Exercise 6: Subset `contrast` to include only the comparisons for the low toxicity compound and create a vector `levels` which can be used to loop through the selected `mean_ID`:**


```{r levels}
# Subsetting the low toxicity (propylene carbonate) data from the contrast dataframe
contrast_low <- contrast %>% 
  filter(file_ID == "CS1104")

contrast_low
```
```{r}
# Contrast_low is separated into concentration levels
levels <- unique(contrast_low$mean_ID)

levels
```

***

Before looping through `levels`, let's go through the steps required to perform a DEG analysis.

**Exercise 7: Using the first value in `levels`, subset the `contrast`, `countData`, and `metaData` to include only the relevant samples:**

<details>
  <summary><strong>Hint</strong></summary>

  Make sure you create new objects, such as `countData_sub` and `metaData_sub` to prevent overwriting the full data frame.

</details>


```{r subset}
# The first low tox concentration level is stored in variable i 
i <- levels[1]
i

```
```{r}
# The first concentration level is subsetted from the contrast_low dataframe
contrast_sub <- contrast_low %>% 
  filter(mean_ID == i)
contrast_sub

# The samples treated with the first concentration level and their controls are subsetted from the metadata
metaData_sub <- metaData %>% 
  filter(mean_ID %in% c(contrast_sub$mean_ID, contrast_sub$mean_ID_control))
metaData_sub

# And from the countdata
countData_sub <- countData[ , metaData_sub$sample_ID]
countData_sub[1:10, ]
```

***

## Fit models

Now that we have subset our data, we can use it to create a DESeq object (`dds`). This also requires us to set our `design`, or model formula. In our data, this will be a simple comparison of two levels of `mean_ID`.

**Exercise 8: Use the `DESeqDataSetFromMatrix()` function to create a `dds` object:**

<details>
  <summary><strong>Hint</strong></summary>

  Designs are a special type of object in R, the formula, which is specified using a tilde (e.g. `~ mean_ID`).

</details>


```{r fit}
# A DESeq object is made with data from the subsetted countdata and metadata
dds <- DESeqDataSetFromMatrix(
  countData = countData_sub,
  colData = metaData_sub,
  # Data will be compared by mean_ID
  design = ~ mean_ID
)
```

***

Next, we use the `DESeq()` function to:

1. Calculate **size factors** using the median ratio method, which normalizes for differences in sequencing depth or library size across samples. The motivation is similar to CPM normalization, but each sample is instead compared to a pseudo-reference sample composed of the geometric mean of all genes. 
2. Estimate **dispersions** (biological variability) for each gene across replicates, in order to properly quantify confidence in results (which will be lower for highly variable genes).
3. Fit a **generalized linear model** (GLM) for each gene, using a negative Binomial distribution. This takes the normalized counts (from size factors), design formula, and estimated dispersion to produce log2 fold changes (log2FC), standard errors (SEs), and p-values.

The output object, `dds`, contains all the model fits and is ready for extracting results.

**Exercise 9: Use the `DESeq()` function to update the `dds` object and include all specified model fits:**

```{r deseq}
# The DESeq function is applied to the DESeq object dds
dds <- DESeq(dds)
dds
```

***

## Extract results

Results can be extracted from a `dds` object using the `results()` function. However, we additionally apply shrinkage to our results. **Shrinkage** pulls extreme or noisy log2FC towards zero when they are not well supported by the data, usually due to:

* Low expression levels
* High dispersion (biological variability)
* Small sample sizes

Shrunk log2FC are more stable and realistic, especially in these scenarios. **Normal** shrinkage uses a Normal prior distribution (bell curve with its peak at 0), which assumes that most genes are not differentially expressed and encourages smaller fold changes unless there is strong evidence to the contrary. 

The final log2FCs are then a posterior estimate that combines:

* The observed data, and
* The prior (Normal distribution centered at 0)

**Exercise 10: Use `lfcShrink()` to extract results for the selected comparison:**

<details>
  <summary><strong>Hint</strong></summary>

  You will get a warning using normal shrinkage, but our experience suggests that it is the best method for these simpler designs.

</details>


```{r shrinkage}
# Applying shrinkage. This way, extreme or noisy values will not have big influences on the data, and the results will be more reliable.
res_low <- lfcShrink(dds, 
                 contrast = c("mean_ID",
                              contrast_sub$mean_ID,
                              contrast_sub$mean_ID_control),
                 type="normal")

head(res_low)
```

***

**Exercise 11: Wrangle the output to additional columns:**

* `gene_symbol` containing the gene names
* `mean_ID` containing the mean ID of the treated samples
* `mean_ID_control` containing the mean ID of the control samples

```{r results}
# res_low is turned into a dataframe
res_low <- data.frame(res_low) %>% 
  # The rownames are turned into a column called gene_symbol
  rownames_to_column(var = "gene_symbol") %>% 
  # Columns containing the mean_ID (which treatment condition the sample was taken from) and the mean_ID_control (which control condition the sample was compared to), as found in contrast_sub, are added. 
  mutate(
    mean_ID = contrast_sub$mean_ID,
    mean_ID_control = contrast_sub$mean_ID_control
  )

head(res_low)
```

***

## Loop

Using a `for` loop, you can run each of these steps for every required comparison.

**Exercise 12: Loop through the remaining values of the `levels` variable to add all of the comparisons for the low toxicity compound to your `res` object:**

```{r loop}
# The stept completed in the previous chunks are repeated with help of a loop for each concentration level other than the first.
for(i in levels[-1]){
  # Creating a contrast dataset and subsetting the metadata and countdata
  contrast_sub <- contrast_low %>% 
    filter(mean_ID == i)
  
  metaData_sub <- metaData %>% 
    filter(mean_ID %in% c(contrast_sub$mean_ID, contrast_sub$mean_ID_control))
  
  countData_sub <- countData[ , metaData_sub$sample_ID]
# Creating a DESeq dataset from the subsetted data
  dds <- DESeqDataSetFromMatrix(
    countData = countData_sub,
    colData = metaData_sub,
    design = ~ mean_ID)  
# The DESeq function is applied  
  dds <- DESeq(dds)
# Normal shrinkage is applied  
  res_low <- rbind(res_low, 
    data.frame(lfcShrink(dds, 
                 contrast = c("mean_ID",
                              contrast_sub$mean_ID,
                              contrast_sub$mean_ID_control),
                 type="normal")) %>% 
      # The data is turned into a clear dataframe, with columns describing the treatment condition and the control condition it was compared to
    rownames_to_column(var = "gene_symbol") %>% 
    mutate(
      mean_ID = contrast_sub$mean_ID,
      mean_ID_control = contrast_sub$mean_ID_control))
}
```
```{r}
# Inspecting the data
head(res_low)
dim(res_low)
```

***

**Exercise 13: Perform DEG analysis for the high toxicity compound, saving the output in `res_high`:**

First, we make the contrast for the high toxicity compound and save the levels:
-> Note: as triazine is a moderately toxic compound, the output will be saved in res_moderate

```{r high-contrast}
# A contrast dataframe is made, containing the treatment conditions / concentration levels for the moderate toxicity compound and the control conditions they are to be compared to
contrast_moderate <- contrast %>% 
  filter(file_ID == "CS1127")

contrast_moderate

levels <- unique(contrast_moderate$mean_ID)

levels
```

Then, we loop through `levels` and save results:

```{r high-loop}
# The function of this loop is the same as the loop for the low toxicity compound, except this time all concentration levels are looped through.
res_moderate <- data.frame()

for(i in levels){
  contrast_sub <- contrast_moderate %>% 
    filter(mean_ID == i)
  
  metaData_sub <- metaData %>% 
    filter(mean_ID %in% c(contrast_sub$mean_ID, contrast_sub$mean_ID_control))
  
  countData_sub <- countData[ , metaData_sub$sample_ID]

  dds <- DESeqDataSetFromMatrix(
    countData = countData_sub,
    colData = metaData_sub,
    design = ~ mean_ID)  
  
  dds <- DESeq(dds)
  
  res_moderate <- rbind(res_moderate, 
    data.frame(lfcShrink(dds, 
                 contrast = c("mean_ID",
                              contrast_sub$mean_ID,
                              contrast_sub$mean_ID_control),
                 type="normal")) %>% 
    rownames_to_column(var = "gene_symbol") %>% 
    mutate(
      mean_ID = contrast_sub$mean_ID,
      mean_ID_control = contrast_sub$mean_ID_control))
}
```
```{r}
# Inspecting the output
head(res_moderate)
dim(res_moderate)
```

***

# Defining DEGs

## Thresholds

There are many ways to define a differentially expressed gene based either on the log2 fold change (log2FC, which reflects the effect size), the adjusted p-value (which reflects the statistical significance after correcting for multiple testing), or a combination of both. For example, some studies may define a gene as differentially expressed if it has an adjusted p-value below 0.05, indicating high statistical confidence, while others may require an absolute log2FC above log2(1.2) (i.e., a twenty percent change) to ensure the change is also biologically meaningful. Combining both criteria helps balance sensitivity and specificity, reducing the chance of identifying genes that are statistically but not biologically relevant, or vice versa.

**Exercise 14: Create two new variables `dir` and `sig` that describe each gene's significance (T/F) and direction of effect (up/down):**

<details>
  <summary><strong>Hint</strong></summary>

  Don't forget to do it for the high and low toxicity compounds.

</details>

```{r sig-dir}

res_low <- res_low %>% 
  mutate(
    # dir describes whether the log2 fold change has gone up (>0) or down (<0) -> the direction (dir) of the change
    dir = ifelse(log2FoldChange < 0, "Down", "Up"),
    # sig describes whether the change is both significant and large enough. If the adjusted p value is equal to or lower than 0.05, AND the absolute log2 fold change is greater than 1.2, the column will return TRUE. Otherwise, it will return FALSE.
    sig = ifelse(padj <= 0.05 & abs(log2FoldChange) > log2(1.2), TRUE, FALSE)
  )

table(res_low$sig, res_low$dir, useNA="always")

res_moderate <- res_moderate %>% 
  mutate(
    dir = ifelse(log2FoldChange < 0, "Down", "Up"),
    sig = ifelse(padj <= 0.05 & abs(log2FoldChange) > log2(1.2), TRUE, FALSE)
  )

table(res_moderate$sig, res_moderate$dir, useNA="always")
```
In the low toxicity compound, there are 21 significant downregulated genes, and 47 significant upregulated genes. In the moderate toxic compound, there are 99 significant downregulated genes and 38 significant upregulated genes. In total, the moderately toxic compound yields more DEGs, though there is (a little) less upregulation and more downregulation.
***

## Save

It is important to save output of analyses locally, especially when they can be quite lengthy to rerun. This will save you considerable time later in the project. 

**Exercise 15: Save the results using the path variables you created at the start of the script:**

```{r save}
save(res_low, file=paste0(root_dir, low_store))
save(res_moderate, file=paste0(root_dir, moderate_store))
```

***

# Plots

## Dose-response plots

One of the plots that is useful to make following DEG analysis is visualization of any dose response effects. In order to ensure that all data point are plotted correctly, it is helpful to create a base data frame for all the direction, concentration combinations.

**Exercise 16: Create a data frame `plot_df` with all compound, concentration, and direction combinations:**

<details>
  <summary><strong>Hint</strong></summary>

  It may be helpful to `rbind()` your high and low toxicity compound results together for visualizations.

</details>

```{r ref-df}
# Moderate and low toxicity compounds are combined into one dataframe, for easier analysis
res <- rbind(res_moderate, res_low)

# A dataframe is created with only the mean_ID and the direction. NA values are filtered out.
plot_df <- res %>% 
  filter(!is.na(dir)) %>% 
  select(mean_ID, dir) %>% 
  distinct()

plot_df
```

***

Using `mean_ID` and `separate()`, you can extract information about the compound and concentration of each experimental group.

**Exercise 17: Create a data frame containing the number of DEGs, concentration, and compound, for each experimental condition and direction:**

<details>
  <summary><strong>Hint</strong></summary>

  Don't forget to set the number of DEGs to 0, where appropriate. 

</details>

```{r plot-df}
# The dataframe created in the previous chunk is adapted.
plot_df <- res %>% 
  # Only genes with significant changes (= DEGs) are taken into account, NA values are filtered out.
  filter(sig == TRUE & !is.na(dir)) %>% 
  # Data in the dataframe is grouped by mean_ID and direction.
  group_by(mean_ID, dir) %>%
  summarize(n = n(), .groups = "drop") %>% 
  # Columns containing mean_ID and direction are added to the dataframe
  right_join(y = plot_df, by = c("mean_ID", "dir")) %>% 
  # If a condition has no DEGs, put a 0 in the "n" column
  mutate(n = ifelse(is.na(n), 0, n)) %>% 
  # Mean_ID is separated into two columns describing the compound and the concentration ID, the mean_ID column itself is not removed. Plate number is removed. 
  separate(mean_ID, sep="_", into=c("Compound", "Conc_ID", NA), remove=F)

plot_df
```

***

**Exercise 18: Plot the number of DEGs across concentrations for the two compounds. Do you see a dose-response?**

```{r dose-response}
plot_df %>% 
  # If a gene is downregulated, show a negative value in the bar plot
  mutate(n = ifelse(dir == "Down", -n, n)) %>% 
  # Bars show the number of DEGs for each concentration level, and are filled by direction
  ggplot(aes(x=Conc_ID, y=n, fill=dir)) +
  geom_col() +
  geom_hline(aes(yintercept=0)) +
  theme_bw() + 
  # Data is separated by compound
  facet_wrap(~Compound, scales="free_x")
```
There seems to be somewhat of a dose-response in the low tox compound: differential gene expression starts out high, and lowers as concentrations increase. No such pattern can be seen in the moderate toxicity compound. There, the lowest concentration also yields the most DEGs. There does seem to be a somewhat consistent transcriptomic response in the higher concentrations, with a larger downregulation at the third and fifth concentration level.

***

## Volcano plots

A volcano plot is useful to visualize effect sizes and significance of each gene in one plot, by plotting log2 fold changes (x-axis) against -log10 p-values (y-axis).

**Exercise 19: Create volcano plots for each of the compounds:**

<details>
  <summary><strong>Hint</strong></summary>

  If you want to plot all concentrations, it may be sensible to create distinct plots for the two compounds. However, you could also plot only the top concentration volcano plot, if the dose response is suggestive that this would be more informative. 

</details>

```{r volcano}
# A plot is created for each unique concentration level
for(i in unique(metaData$compound_ID)[1:2]) {
  print(res %>% 
          # mean_ID is separated into compound and concentration ID, plate number is removed.
          separate(mean_ID, sep="_", into=c("Compound", "Conc_ID", NA), remove=F) %>% 
          # NA values are filtered out, output is separated by compound
          filter(!is.na(sig) & Compound == i) %>% 
          # A variable is added to calculate the -log10 of the p value
          mutate(log10p = -log10(pvalue)) %>% 
          ggplot(aes(x = log2FoldChange, y = log10p)) +
          # Significant genes are colored
          geom_point(aes(color=sig), size=0.5) +
          # Labels with gene symbol are added to DEGs that have a very low p value
          geom_text_repel(aes(label = ifelse(padj < 1E-4, gene_symbol, NA)),
                    color="dodgerblue", size=2) +
          # Minimum and maximum x-axis values are set to the absolute maximum fold change, with NA values removed
          xlim(-max(abs(res$log2FoldChange), na.rm=T), max(abs(res$log2FoldChange), na.rm=T)) +
          scale_color_manual(values = c("grey", "dodgerblue")) +
          theme_bw() + 
          # Each plot is titled with the compound ID
          ggtitle(i) +
          # A separate plot is made for each concentration level
          facet_wrap(~Conc_ID))
}
```
In the first two concentration levels of the low tox compound, the same few genes have the highest differential expression. These genes are:
- PAEP - Involved in pregnancy
- MYH11 - Smooth muscle myosin
- S100A8 - Regulation of cellullar processes such as cell cycle progression and differentiation
- DCN - Collagen fibril assembly, tumor suppression -> stimulation of autophagy and inflammation

These genes, except maybe for DCN, do not make a lot of sense in a toxicity context.

In the cells treated with a moderately toxic chemical, the genes with the largest differential expression vary per concentration level, although APOA4 can be seen in concentration level 2, 3, and 4. The genes with the largest differential expression are:
- RPL27 - Ribosomal protein
- DLG1 - Scaffold protein. Plays a role in septate junction formation, signal transduction, cell proliferation, synaptogenesis and lymphocyte activation.
- APOA4 - Exact function unknown. Possibly an activator of lecithin-cholesterol acyltransferase.
- ALDOB - Conversion of fructose-1,6-bisphosphate to glyceraldehyde 3-phosphate and dihydroxyacetone phosphate. Biased expression in the liver.
- TRMT112 - Macromolecule methylation, macromolecule metabolic processes

These genes, too, do not make a lot of sense in a toxicity context, although activation of DLG1 could be a sign of toxicity (heightened signal transduction, lymphocyte activation).
***

## MA plots

One feature that volcano plots miss is the baseline expression of genes. This can mean that very lowly expressed genes appear important (due to high significance and effect sizes). Visualizing results using an MA plot, which plots log2 base mean (y-axis) against log2 fold changes (x-axis), can be helpful to distinguish between noisy genes and biologically relevant ones.

**Exercise 20: Create MA plots for the two compounds. Outline your findings:**

```{r ma}
# A plot is created for each unique treatment condition.
for(i in unique(metaData$compound_ID)[1:2]) {
  print(res %>% 
          separate(mean_ID, sep="_", into=c("Compound", "Conc_ID", NA), remove=F) %>% 
          filter(!is.na(sig) & Compound == i) %>% 
          # Log2 of the base mean (mean of normalised count of all samples) is used as the x-axis
          mutate(log2baseMean = log2(baseMean)) %>% 
          ggplot(aes(x = log2baseMean, y = log2FoldChange)) +
          geom_point(aes(color=sig), size=0.5) +
          geom_text_repel(aes(label = ifelse(padj < 1E-4, gene_symbol, NA)),
                    color="dodgerblue", size=2) +
          ylim(-max(abs(res$log2FoldChange), na.rm=T), max(abs(res$log2FoldChange), na.rm=T)) +
          scale_color_manual(values = c("grey", "dodgerblue")) +
          theme_bw() + 
          ggtitle(i) +
          facet_wrap(~Conc_ID))
}
```
The MA plot for the low toxic compound shows the same four genes as being the most differently expressed. 
For the moderately toxic compound, RPL27 does not show anymore. 
***

# Advanced question

## Gene level plots

It can also be useful to plot individual gene results, especially to investigate gene-level dose responses.

**Exercise 21: Plot the top 16 genes for each of the compounds in a dose-response plot. Are there any genes with particular relevance to hepatotoxicity?**

```{r high-gene}
#

```

***

# Session Info

```{r session-info}
#
devtools::session_info()
```

***

This is the end of the differential gene expression analysis. Next, please move on to functionally interpret your results using over-representation analysis in `06_ORA_Outline.Rmd`. Example answers will be available from the GitHub at EOB on Friday.

***

