---
title: "BOO 2025"
subtitle: "Script 3: Sample QC"
date: "`r Sys.Date()`" 
author: 
  Adam Seuren
output:
  html_document:
    code_download: true
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

***

> In this script, you will perform sample-level quality control (QC), removing any poor quality samples and ensuring that experimental replicates are comparable to one another. 

***

# Setup

## Clean

As before, we perform several setup steps at the start of the script to ensure our work is reproducible and clear. 

**Exercise 1: Download the R markdown template and clean your environment:**

```{r clean}
# 
rm(list=ls())
```

***

## Set variables

**Exercise 2: Create the following objects in your R environment:**

* `root_dir` - project folder
* `count_path` - location of the `countData` object within the project folder
* `cpm_path` - location of the `cpmData` object
* `metadata_path` - location of the `metaData` object

* `count_store` - location to save the `countData` object within the project folder
* `cpm_store` - location to save the `cpmData` object
* `metadata_store` - location to save the filtered `metaData` after QC

* `count_threshold` - minimum library size required for samples (600,000; 20% of the target sequencing depth)
* `corr_threshold` - required correlation between replicates (0.9)

```{r set-variables}
#
root_dir <- "C:/Users/mmpse/Documents/GitHub/BOO-Adam/QC/03_Sample_QC/" 

count_path <- "Input/countData.Rdata" 
cpm_path <- "Input/cpmData.Rdata" 
metadata_path <- "Input/metaData.Rdata"

count_store <- "Output/countData.Rdata"
cpm_store <- "Output/cpmData.Rdata"
metadata_store <- "Output/metaData.Rdata" 

# Set count threshold to 600,000 - this is the minimum library size a sample needs to have. Including smaller samples would provide more unreliable data, as a small change in gene expression might look like a large effect.
count_threshold <- 6E5

# Set correlation threshold to 0.9 - it is important for the quality of the analysis that replicates of the same condition are correlated with each other
corr_threshold <- 0.9 
```

***

## Packages

Here, we load `tidyverse` and also a new package:

* `ggrepel` allows us labels in plots to "repel" each other and make visualizations clearer

**Exercise 3: Load `tidyverse` and `ggrepel` into your environment:**

```{r load-packages, warning=F, message=F}
#
library(tidyverse)
# ggrepel gives the possibility for labels in plots to stop overlapping with each other
library(ggrepel)
```

***

## Load data

**Exercise 4: Load the count data, CPM data, and metadata into your environment:**

<details>
  <summary><strong>Hint</strong></summary>

  Make sure these are the ones your saved at the end of the probe QC.

</details>

```{r load-data, warning=FALSE, message=FALSE}
#
load(paste0(root_dir, count_path))
load(paste0(root_dir, cpm_path))
load(paste0(root_dir, metadata_path))
```

***

# Library size

## Check

Before applying any filters, it is good to perform some checks.

**Exercise 5: Check that the column names of `countData` match the `sample_ID` order in `metaData`:**

```{r order-check}
# 
table(metaData$sample_ID == colnames(countData))
```
The function returns TRUE, so the column names of countData match the order of sample_ID in the metadata.
***

## Calculate

**Exercise 6: Now that we have removed unreliable and lowly expressed probes from `countData`, recalculate and save a new `lib_size` in the metadata:**

```{r calculate-lib}
# Summary of the unchanged lib_size
summary(metaData$lib_size)
```
Mean library size: 3740053
```{r}
# setting the library size as the total counts per sample of the updated countData
metaData$lib_size <- colSums(countData)

summary(metaData$lib_size)
```
Mean library size: 3614091. This is a bit smaller than before, as unreliable and lowly expressed probes have been removed from the countData.  
***

## Distribution

**Exercise 7: Make a histogram of `lib_size`. What range of values does this variable take and is this reasonable for a TempO-Seq experiment?**

```{r lib-histogram}
#
metaData %>% 
  # Creating a histogram of library size
  ggplot(aes(x = lib_size)) +
  geom_histogram(fill = "lightblue", color = "black", bins = 30) +
  ggtitle("Histogram of library size values") + 
  xlab("Library size") +
  theme_bw()
```

The library size ranges from a bit over 2 million to a bit over 5 million reads, with an average around 3 million. This is a good amount for a TempOseq analysis, as the method is highly efficient.
***

## Flag

Samples whose library size is below 20% of the targeted sequencing depth (`corr_threshold`; 600,000) should be flagged as having low reads.

**Exercise 8: Create a flag in `metaData` describing if samples have low reads or not:**

```{r lib-flag}
metaData <- metaData %>% 
  # A column called flagLowReads is added to the metadata, which tells the reader whether or not the sample library size is above the threshold.
  mutate(flagLowReads = ifelse(lib_size <= count_threshold, T, F))

table(metaData$flagLowReads)
```
All values in the new column return false, which means that no sample library size is below the count threshold of 600000. 
***

## Plot

It is good to visualize the library size for each sample, grouped by compound ID. This shows us whether samples are well above the threshold, and allows us to inspect the data more carefully.

**Exercise 9: Create a boxplot for the library sizes of each compound (including DMSO) and describe any patterns you identify:**

<details>
  <summary><strong>Hint</strong></summary>

  You can colour your boxplots by concentration to visualize patterns more clearly.

</details>

```{r lib-boxplot}
#
metaData %>%   
  # concentration ID is made into a factor, so it can be used to color the boxplot by
  mutate(conc_ID = factor(conc_ID)) %>%   
  # library size is plotted against compound ID; we want to view the library size per condition
  ggplot(aes(x=compound_ID, y=lib_size)) + 
  # The plot is colored by concentration
  geom_boxplot(aes(color=conc_ID), width=0.8) +  
  # A dashed line is drawn at the minimum library size of 600000 counts.
  geom_hline(aes(yintercept=count_threshold), color="black", linetype="dashed") +
  # values below the threshold are labelled
  geom_text_repel(aes(x = compound_ID, y = lib_size, color = conc_ID),   
                   label=ifelse(metaData$lib_size < count_threshold, 
                                metaData$rep, "")) +

  xlab("") + ylab("Library size") + ggtitle("Library size distributions per condition") +    
  theme_bw()                                  
```

The library sizes per condition are relatively comparable to each other. Library sizes in the highest concentrations of both treatment conditions are smaller than the library sizes of the conditions of the lower concentrations; this could point to cytotoxicity, as dying cells do not express es much DNA. Additionally, the library size in the cells treated with propylene carbonate (the low-toxic chemical) ascends over the concentration levels, whereas the library size in the cells treated with triazine (the medium-toxic chemical) is consistently higher. This could be a sign that triazine is, indeed, more toxic (as toxicity can increase gene expression and therefore a larger library size), but that propylene carbonate, too, causes reactions in the cell when the concentration is high enough.
***

# Replicate correlation

## log2CPM

The replicate correlation filter aims to remove any outlying replicates, with maximum pairwise correlations below the `corr_threshold` (set to 0.9). We usually perform this correlation analysis on the log2CPM values to ensure highly expressed genes do not have undue influence on the correlation values. A value of 1 is added to the CPM, to prevent issues arising from the fact that `log2(0)` is `-Inf`. 

**Exercise 10: Calculate and store the log2(CPM + 1) values in a `logcpmData` object:**

```{r log2cpm}
# the log2(cpm + 1) is calculated and stored
logcpmData <- log2(cpmData + 1)
```

***

## Pairwise correlations

In order to calculate pairwise correlations, each sample needs to be compared to the other replicates in its experimental group. We can do this by looping through `mean_ID`.

**Exercise 11: Calculate the pairwise replicate correlations for this data:**

<details>
  <summary><strong>Hint</strong></summary>

  The correlation can be calculated using `cor(cpmDataReps[,j], cpmDataReps[,k])` within an appropriate loop.

</details>

```{r pairwise-corr}
# A data frame is made for the output of the replicate filter
replicateFilterOutput <- data.frame()

# A loop is made with an iteration for each unique condition
for(i in unique(metaData$mean_ID)){
  # The metadata is filtered to only show data from the current mean_id
  metaDataReps <- metaData %>% 
    filter(mean_ID == i)
  
  # The repetitions from the current mean_id are subsetted out of the logcpmdata
  cpmDataReps <- logcpmData[, metaDataReps$sample_ID] 
  
  # loop through each column of the cpmdata rows that were selected in the previous line   
  for(j in 1:ncol(cpmDataReps)){
    # in a pairwise fashion (comparing the replicates with each other in pairs)
    for(k in 1:ncol(cpmDataReps)){
      # save the two selected replicates as variables
      sample_A <- colnames(cpmDataReps)[j]
      sample_B <- colnames(cpmDataReps)[k]
      
      # the correlation between identical replicates does not need to be calculated (do not compare a sample with itself)
      if(sample_A != sample_B){
        # calculate pairwise correlation and put it into a variable called r2
        r2 <- cor(cpmDataReps[,j], cpmDataReps[,k])
        
        # put the results of the calculation into the replicateFilterOutput dataframe
        replicateFilterOutput <- rbind(
          replicateFilterOutput, 
          data.frame(
            # first column: currently selected mean_ID
            mean_ID = i, 
            # second column: sample_ID of sample A
                     sample_A = sample_A,
            # third column: sample_ID of sample B
                     sample_B = sample_B,
            # fourth column: correlation between the two samples
                     r2 = r2))
      }
    }
  }
}

# show the first rows of the dataframe
head(replicateFilterOutput)

```

***

## Maximum

Each sample is judged by the best pairwise correlation it can achieve. If this is below `corr_threshold`, the sample should be flagged.

**Exercise 12: Calculate the `max_r2` for each sample and add it to the `replicateFilterOutput`:**

```{r max-r2}
replicateFilterOutput <- replicateFilterOutput %>% 
  # the sample name is changed to compound name _ concentration ID (plate and replicate number are removed from the name)
  separate(sample_A, 
           into = c("Compound", "Conc_ID", NA, NA), 
           remove = F, 
           sep = "_") %>% 
  # compound column is added. 
  mutate(Compound = 
           # if the compound name is DMSO
           ifelse(grepl("DMSO", Compound), 
                  # select only the first 5 letters (in order to not get separations between the samples in the same DMSO level; names in the column are DMSOH and DMSOL instead of DMSOHigh1 and DMSOHigh2, DMSOLow1 and DMSOLow2)
                  substr(Compound,1,5), Compound)) %>% 
  # group by condition
  group_by(sample_A) %>%
  # save the maximum pairwise correlation between replicates of that condition
  mutate(max_r2 = max(r2, na.rm = T)) %>% 
  # ungroup the dataframe in order to avoid errors later on
  ungroup()

# inspect the maximum correlations between replicates
summary(replicateFilterOutput$max_r2)

```
The lowest of the maximum pairwise correlations between replicates is 0.9071, which is above the threshold of 0.9.
***

## Plot

**Exercise 13: Visualize the pairwise replicate correlations for each experimental conditions. Describe what you observe:**

```{r corr-boxplot}
replicateFilterOutput %>% 
  # pairwise correlation between replicates is plotted into a boxplot per condition
  ggplot(aes(x = sample_A, y = r2)) +
  geom_boxplot(color = "grey60") +
  # with points at the actual pairwise correlation values
  geom_point(color = "black", size = 0.5) +
  # where the point describing the highest pairwise correlation level for each condition is colored, with a different color for eash concentration level
  geom_point(aes(y = max_r2, color = Conc_ID), 
             size = 1.5) +
  # a dashed line is added at the correlation threshold value of 0.9
  geom_hline(aes(yintercept = corr_threshold), 
             color = "red", linetype = "dashed") +
  ylab("") + xlab("Sample ID") + ggtitle("Replicate correlations") +
  theme_bw() +
  # sample names are not printed (they are not readable when printed, and are not necessary for the clarity of the plot)
  theme(axis.text.x = element_blank()) +
  # each compound gets a different plot, with an x-axis that fits the plot
  facet_wrap(~Compound, scales='free_x', nrow=2)

```

Correlation levels are generally high. However, the correlations of one sample treated with triazine at concentration level 5 is noticably lower than the rest. It is close to the threshold, though still above it. All samples pass the correlation filter.
***

## Flag

**Exercise 14: Flag any samples that did not pass the replicate correlation filter in the `metaData`:**

<details>
  <summary><strong>Hint</strong></summary>

  You can merge the replicate correlation filter output with the metaData to create a `max_r2` column after some processing.

</details>

```{r corr-flag}
# the maximum correlation between replicates is filtered for each condition and saved into a new variable
replicateFilterMerge <- replicateFilterOutput %>% 
  select(sample_ID = sample_A, max_r2) %>% 
  # for each unique condition
  distinct()

# the new variable is added to the metadata file, by sample ID -> the metadata now shows the maximum correlation level for each sample.
metaData <- left_join(metaData, replicateFilterMerge, 
                      by = "sample_ID") %>% 
  # a new column is added, which shows if the maximum correlation levels are above the threshold value of 0.9.
  mutate(flagLowCorr = ifelse(max_r2 <= corr_threshold, T, F))

table(metaData$flagLowCorr)

```
All samples pass the correlation filter, no samples are flagged.
***

# Advanced questions

If you would like a bit more of a challenge, here are a few extra questions relating to the two sample QC steps above. However, you can also skip these, save your data, and move on to the PCA.

## Library size

**Exercise 14: What are the benefits of a sample having a higher library size and does this benefit apply to some genes more than others?**

```{r read-depth}
#

```

***

## Replicate correlation

Instead of looking at pairwise correlations, another way of measuring how good a replicate is is by comparing it to the average for that experimental condition. 

**Exercise 15: Calculate replicate correlation in this way and see if it alters the results of this filter. What is one benefit and downside of assessing replicate correlation in this manner?**

```{r mean-corr}
#

```

***

# Save

**Exercise 16: Remove samples that did not pass the sample QC steps from your data:**

<details>
  <summary><strong>Hint</strong></summary>

  Don't forget to also subset the count and CPM data.

</details>

```{r any-flag}
# metadata is subsetted to only show the samples that aren't flagged. this is only the high quality data.
metaData <- metaData %>% 
  filter(!flagLowReads & !flagLowCorr)

# countdata and cpmdata are subsetted to only show the samples that are in the updated metadata file
cpmData <- cpmData[ , metaData$sample_ID]
countData <- countData[ , metaData$sample_ID]

# dimensions of the new metadata file are checked (should be 66/32 because no samples were flagged, so no rows should be removed)
dim(metaData)
```

***

## Save

**Exercise 17: Save the updated data:**

```{r save-metadata}
save(metaData, file=paste0(root_dir, metadata_store))
save(cpmData, file=paste0(root_dir, cpm_store))
save(countData, file=paste0(root_dir, count_store))
```

***

# Session Info

**Exercise 18: Print your session info at the end of the script, knit the R markdown document, and push it to GitHub:**

```{r session-info}
devtools::session_info()
```

***

That is the end of the Sample QC. Example answers will be available from the `BOO_template` GitHub on Tuesday. 

Next, please move on to the PCA using `04_PCA_Outline.Rmd`.

***

